src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1360,24 +1483,62 @@ CHIP_ERROR CASESession::HandleSigma3(System::PacketBufferHandle && msg)
         P256PublicKeyHSM initiatorPublicKeyHSM;
         memcpy(Uint8::to_uchar(initiatorPublicKeyHSM), initiatorPublicKey.Bytes(), initiatorPublicKey.Length());
         SuccessOrExit(
-            err = initiatorPublicKeyHSM.ECDSA_validate_msg_signature(msg_R3_Signed.Get(), msg_r3_signed_len, tbsData3Signature));
+            err = initiatorPublicKeyHSM.ECDSA_validate_msg_signature(work.msg_R3_Signed.Get(), work.msg_r3_signed_len, work.tbsData3Signature));
     }
 #else
-    SuccessOrExit(err = initiatorPublicKey.ECDSA_validate_msg_signature(msg_R3_Signed.Get(), msg_r3_signed_len, tbsData3Signature));
+    SuccessOrExit(err = initiatorPublicKey.ECDSA_validate_msg_signature(work.msg_R3_Signed.Get(), work.msg_r3_signed_len, work.tbsData3Signature));
 #endif
 
-    SuccessOrExit(err = mCommissioningHash.Finish(messageDigestSpan));
+exit:
+    work.status = err;
+
+    auto err2 = DeviceLayer::PlatformMgr().ScheduleWork(
+        [](intptr_t arg) {
+            auto & work2 = *reinterpret_cast<Sigma3Work*>(arg);
+            work2.session->HandleSigma3c(work2);
+        },
+        reinterpret_cast<intptr_t>(&work));
+
+    if (err2 != CHIP_NO_ERROR)
+    {
+        delete &work; // scheduling failed, so delete
+    }
+}
+
+CHIP_ERROR CASESession::HandleSigma3c(Sigma3Work & work)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    bool ignoreFailure = true;
+
+    // Special case: if for whatever reason not in expected state or sequence,
+    // don't do anything, including sending a status report or aborting the
+    // pending establish.
```

Comment:
How could this happen?

Reply:
Shouldn't normally happen, but let's say you start connecting, background work is starved, in the meantime the pending establish is stopped and another one starts, then finally the background work proceeds. We'd be here, but sequence or state wouldn't match, so we'd just not continue doing the work requested (but not interfere with any new work).

Reply:
Suggest describing that in the comment